lib/src/rules/weak_mode_unnecessary_null_checks.dart

```diff
@@ -0,0 +1,104 @@
+// Copyright (c) 2020, the Dart project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:analyzer/dart/analysis/features.dart';
+import 'package:analyzer/dart/ast/ast.dart';
+import 'package:analyzer/dart/ast/token.dart';
+import 'package:analyzer/dart/ast/visitor.dart';
+
+import '../analyzer.dart';
+
+const _desc = r'Unnecessary null check for non-nullable value.';
+
+const _details = r'''
+
+Unnecessary null check for non-nullable value in a opted-out library.
+
+**BAD:**
+```
+nonNullable?.property;
+if (nonNullable != null) {}
+```
+
+**GOOD:**
+```
+nonNullable.property;
+```
+
+''';
+
+class WeakModeUnnecessaryNullChecks extends LintRule implements NodeLintRule {
+  WeakModeUnnecessaryNullChecks()
+      : super(
+            name: 'weak_mode_unnecessary_null_checks',
+            description: _desc,
+            details: _details,
+            group: Group.style);
+
+  @override
+  void registerNodeProcessors(
+      NodeLintRegistry registry, LinterContext context) {
+    final visitor = _Visitor(this, context);
+    registry.addCompilationUnit(this, visitor);
+  }
+}
+
+class _Visitor extends RecursiveAstVisitor<void> {
+  _Visitor(this.rule, this.context);
+
+  final LintRule rule;
+  final LinterContext context;
+
+  @override
+  void visitCompilationUnit(CompilationUnit node) {
+    if (node.featureSet.isEnabled(Feature.non_nullable)) {
```

Comment:
The current convention is to use `LinterContext.isEnabled` to test this, and to put the test in `registerNodeProcessors` if the whole rule is conditional. By not registering the visitor at all in those cases we improve overall performance.